File: scratch/test_websocket.py
import websocket
import json
import threading
import time
import sys

anonKey = "your-anon-key"
merchantId = "your-merchant-uuid"
wsURLString = f"wss://your-project-ref.supabase.co/realtime/v1/websocket?apikey={anonKey}&vsn=1.0.0"

def on_message(ws, message):
    try:
        data = json.loads(message)
        event = data.get("event")
        topic = data.get("topic")
        payload = data.get("payload", {})
        
        # Print event details
        print(f"\n[WS MESSAGE] Topic: {topic}, Event: {event}", flush=True)
        if event == "postgres_changes":
            data_payload = payload.get("data", {})
            table = data_payload.get("table")
            type_ = data_payload.get("type")
            record = data_payload.get("record") or data_payload.get("old_record") or {}
            print(f"  Table: {table}, Type: {type_}", flush=True)
            print(f"  Record: {json.dumps(record, indent=2)}", flush=True)
        else:
            print(f"  Payload: {json.dumps(payload, indent=2)}", flush=True)
    except Exception as e:
        print(f"Error parsing message: {e}\nRaw: {message}", flush=True)

def on_error(ws, error):
    print(f"WebSocket Error: {error}", flush=True)

def on_close(ws, close_status_code, close_msg):
    print(f"WebSocket Closed: status={close_status_code}, msg={close_msg}", flush=True)

def on_open(ws):
    print("WebSocket Connected! Sending join payload...", flush=True)
    
    join_payload = {
        "topic": "realtime:public",
        "event": "phx_join",
        "payload": {
            "config": {
                "postgres_changes": [
                    {"event": "*", "schema": "public", "table": "orders", "filter": f"merchant_id=eq.{merchantId}"},
                    {"event": "*", "schema": "public", "table": "order_items", "filter": f"merchant_id=eq.{merchantId}"},
                    {"event": "*", "schema": "public", "table": "service_requests", "filter": f"merchant_id=eq.{merchantId}"}
                ]
            }
        },
        "ref": "1"
    }
    
    ws.send(json.dumps(join_payload))
    
    # Heartbeat thread
    def send_heartbeat():
        while True:
            time.sleep(20)
            try:
                heartbeat = {
                    "topic": "phoenix",
                    "event": "heartbeat",
                    "payload": {},
                    "ref": "heartbeat"
                }
                ws.send(json.dumps(heartbeat))
                print("[WS SENT] Heartbeat", flush=True)
            except Exception as e:
                print(f"Failed to send heartbeat: {e}", flush=True)
                break
                
    t = threading.Thread(target=send_heartbeat, daemon=True)
    t.start()

if __name__ == "__main__":
    print(f"Connecting to: {wsURLString}", flush=True)
    websocket.enableTrace(False)
    ws = websocket.WebSocketApp(
        wsURLString,
        on_open=on_open,
        on_message=on_message,
        on_error=on_error,
        on_close=on_close
    )
    ws.run_forever()

File: scratch/test_websocket.swift
import Foundation

let environment = ProcessInfo.processInfo.environment
guard let anonKey = environment["SUPABASE_ANON_KEY"] else {
    fatalError("SUPABASE_ANON_KEY is required")
}
let merchantId = "your-merchant-uuid"

let supabaseURL = environment["SUPABASE_URL"] ?? "http://119.59.99.163"
guard var components = URLComponents(string: supabaseURL) else {
    fatalError("Invalid SUPABASE_URL")
}
components.scheme = components.scheme == "https" ? "wss" : "ws"
components.path = "/realtime/v1/websocket"
components.queryItems = [
    URLQueryItem(name: "apikey", value: anonKey),
    URLQueryItem(name: "vsn", value: "1.0.0")
]
let wsURLString = components.string ?? ""
guard let url = URL(string: wsURLString) else {
    print("Invalid URL")
    exit(1)
}

print("Connecting to \(url.absoluteString)...")
let session = URLSession(configuration: .default)
let task = session.webSocketTask(with: url)
task.resume()

func listen() {
    task.receive { result in
        switch result {
        case .success(let message):
            switch message {
            case .string(let text):
                print("\n[WS RECEIVED STRING]:\n\(text)")
            case .data(let data):
                if let text = String(data: data, encoding: .utf8) {
                    print("\n[WS RECEIVED DATA]:\n\(text)")
                } else {
                    print("\n[WS RECEIVED BINARY DATA]: \(data.count) bytes")
                }
            @unknown default:
                break
            }
            listen()
        case .failure(let error):
            print("WebSocket error: \(error)")
            exit(1)
        }
    }
}

// Start listening
listen()

// Send Join Payload
let joinPayload: [String: Any] = [
    "topic": "realtime:public",
    "event": "phx_join",
    "payload": [
        "config": [
            "postgres_changes": [
                ["event": "*", "schema": "public", "table": "orders", "filter": "merchant_id=eq.\(merchantId)"],
                ["event": "*", "schema": "public", "table": "order_items", "filter": "merchant_id=eq.\(merchantId)"],
                ["event": "*", "schema": "public", "table": "service_requests", "filter": "merchant_id=eq.\(merchantId)"]
            ]
        ]
    ],
    "ref": "1"
]

if let data = try? JSONSerialization.data(withJSONObject: joinPayload, options: []),
   let jsonString = String(data: data, encoding: .utf8) {
    print("Sending join payload...")
    task.send(.string(jsonString)) { error in
        if let error = error {
            print("Failed to send join payload: \(error)")
        } else {
            print("Join payload sent successfully!")
        }
    }
}

// Start heartbeat timer
let timer = Timer.scheduledTimer(withTimeInterval: 20.0, repeats: true) { _ in
    let heartbeat: [String: Any] = [
        "topic": "phoenix",
        "event": "heartbeat",
        "payload": [:],
        "ref": "heartbeat"
    ]
    if let data = try? JSONSerialization.data(withJSONObject: heartbeat, options: []),
       let jsonString = String(data: data, encoding: .utf8) {
        task.send(.string(jsonString)) { error in
            if let error = error {
                print("Heartbeat failed: \(error)")
            } else {
                print("Heartbeat sent.")
            }
        }
    }
}

// Run the runloop to keep script alive
RunLoop.main.run()
